pkg/vm/builtins/ecop.go

```diff
@@ -0,0 +1,234 @@
+	r, err := ecop(&p, &q, inputsFelt[4], &alpha)
+	if err != nil {
+		return err
+	}
+
+	// store the resulting point `r`
+	outputOff := inputOff + inputCellsPerEcOp
+
+	rxMV := mem.MemoryValueFromFieldElement(&r.X)
+	err = segment.Write(outputOff, &rxMV)
+	if err != nil {
+		return err
+	}
+
+	ryMV := mem.MemoryValueFromFieldElement(&r.Y)
+	err = segment.Write(outputOff+1, &ryMV)
+	return err
+}
+
+// structure to represent a point in the elliptic curve
+type point struct {
+	X, Y f.Element
+}
+
+// returns true if a point `p` belongs to the `ec` curve ruled by the params `alpha` and
+// `beta`. In other words, true if  y^2 = x^3 + alpha * x + beta
+func (p *point) onCurve(alpha, beta *f.Element) bool {
+	// calculate lhs
+	y2 := f.Element{}
+	y2.Square(&p.Y)
+
+	// calculate rhs
+	x3 := f.Element{}
+	x3.Square(&p.X)
+	x3.Mul(&x3, &p.X)
+
+	ax := f.Element{}
+	ax.Mul(alpha, &p.X)
+
+	x3.Add(&x3, &ax)
+	x3.Add(&x3, beta)
+
+	// return lhs == rhs
+	return y2.Equal(&x3)
+}
+
+// returns the result of the ecop operation on points `P` and `Q` with scalar
+// `m` and param `alpha`. The resulting point `R` is equal to  P + m * Q
+func ecop(p *point, q *point, m, alpha *f.Element) (point, error) {
+	partialSum := *p
+	doublePoint := *q
+
+	mBytes := m.Bytes()
+	scalar := uint256.Int{}
+	scalar.SetBytes32(mBytes[:])
+
+	// Note: In the python VM the height is a parameter but it is always set at 256
+	// therefore we treat it as a constant
+	const height = 256
+	// todo(rodro): iteration could be cut short on the biggest bit with a one of the `scalar`
+	for i := 0; i < height && !scalar.IsZero(); i++ {
+		// we check that both points are always different between each others
+		// `ecadd` assume `x` ordinates are always different
+		// `ecdouble` assumes `y` coordinates are always different
+		if doublePoint.X.Equal(&partialSum.X) || doublePoint.Y.Equal(&partialSum.Y) {
+			return point{}, fmt.Errorf(
+				"at least one similar coordinate between P(%s, %s) and Q(%s, %s)",
+				&p.X, &p.Y, &q.X, &q.Y,
+			)
+		}
+		and := uint256.Int{}
+		and.And(&scalar, &utils.Uint256One)
+		if !and.IsZero() {
+			partialSum = ecadd(&partialSum, &doublePoint)
+		}
+
+		// todo(rodro): This loop can be optimized, potentially innecesary shift operations
+		doublePoint = ecdouble(&doublePoint, alpha)
+		scalar.Rsh(&scalar, 1)
+	}
+
+	return partialSum, nil
+}
+
+// performs elliptic curve addition over two points. Assumes `x` ordinates are
+// always different
+func ecadd(p *point, q *point) point {
+	// get the slope between the two points
+	slope := f.Element{}
+	slope.Sub(&p.Y, &q.Y)
+	denom := f.Element{}
+	denom.Sub(&p.X, &q.X)
+	slope.Div(&slope, &denom)
+
+	// get the x coordinate: x = slope^2 - p.X - q.X
+	x := f.Element{}
+	x.Square(&slope)
+	x.Sub(&x, &p.X)
+	x.Sub(&x, &q.X)
+
+	// get the y coordinate: y = slope * (p.X - x) - p.Y
+	y := f.Element{}
+	y.Sub(&p.X, &x)
+	y.Mul(&y, &slope)
+	y.Sub(&y, &p.Y)
+
+	return point{x, y}
+}
+
```

Comment:
Not sure if the comment is correct here. Isn't here only one `y` coordinate here?